common/aws/s3/client.go

```diff
@@ -148,3 +175,162 @@ func (s *client) ListObjects(ctx context.Context, bucket string, prefix string)
 	}
 	return objects, nil
 }
+
+func (s *client) CreateBucket(ctx context.Context, bucket string) error {
+	_, err := s.s3Client.CreateBucket(ctx, &s3.CreateBucketInput{
+		Bucket: aws.String(bucket),
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (s *client) FragmentedUploadObject(
+	ctx context.Context,
+	bucket string,
+	key string,
+	data []byte,
+	fragmentSize int) error {
+
+	fragments, err := BreakIntoFragments(key, data, s.cfg.FragmentPrefixChars, fragmentSize)
+	if err != nil {
+		return err
+	}
+	resultChannel := make(chan error, len(fragments))
+
+	ctx, cancel := context.WithTimeout(ctx, s.cfg.FragmentWriteTimeout)
+	defer cancel()
+
+	for _, fragment := range fragments {
+		fragmentCapture := fragment
+		s.pool.Submit(func() {
```

Comment:
This pool is shared across all ongoing uploads and downloads. Which means when there are many uploads/downloads in flight, it can build backpressure. Is that problematic?

Reply:
My previous iteration had a config parameter that allowed the client to be have a configurable sized work queue (i.e. when I was using my own implementation of `workerpool`. Unfortunately, the `workerpool` library does not allow us to override the channel used to send data to the workers.

From the implementation:
```
workerQueue: make(chan func())
```

By default, channels have a buffer size of 0. This means that if the number of read/write tasks exceed the number of available workers, the caller will block until all tasks are accepted by a worker. This will provide back pressure if more work is scheduled than there are workers to handle that work.

Are you ok with the way this is configured by default? If not, we will probably need our own implementation of `workerpool`.